File: Data Structure & Algorithm/Template/BFS & DFS/BFS/BFS.py
from typing import Any, List


class TestNode:

    def __init__(self, value: Any = 0, neighbors: List = []):
        self.value: Any = value
        self.neighbors: List[TestNode] = neighbors


def BFS(root: TestNode) -> int:
    queue: List[TestNode] = [root]
    layer = 0

    while len(queue) > 0:
        size = len(queue)
        for _ in range(size):
            current = queue[0]
            if current:
                print(current.value)
            if len(current.neighbors) > 0:
                for item in current.neighbors:
                    queue.append(item)
            queue.pop(0)
        layer = layer + 1

    return layer


# test
root = TestNode(1, [])
node2 = TestNode(2, [])
node3 = TestNode(3, [])
node4 = TestNode(4, [])
node5 = TestNode(5, [])
node6 = TestNode(6, [])
node7 = TestNode(7, [])

root.neighbors.append(node2)
root.neighbors.append(node3)
node2.neighbors.append(node4)
node2.neighbors.append(node5)
node3.neighbors.append(node6)
node6.neighbors.append(node7)

res = BFS(root)
print(f"layer: {res}")

File: Data Structure & Algorithm/Template/BFS & DFS/BFS/BFS.ts
interface TestNode {
  value: any;
  neighbors: TestNode[];
}

export function BFS(root: TestNode): number {
  // store all nodes which are waiting to be processed
  const queue: TestNode[] = [root];
  // number of steps needed from root to current node
  let layer = 0;
  // BFS
  while (queue.length) {
    // iterate the nodes which are already in the queue
    const size = queue.length;
    for (let i = 0; i < size; i++) {
      const current = queue.shift()!;
      if (current) {
        console.log(current.value);
      }
      for (const next of current.neighbors) {
        queue.push(next);
      }
    }
    layer = layer + 1;
  }
  return layer;
}

// test
const root: TestNode = {
  value: 1,
  neighbors: [
    {
      value: 2,
      neighbors: [
        {
          value: 4,
          neighbors: [],
        },
        {
          value: 5,
          neighbors: [],
        },
      ],
    },
    {
      value: 3,
      neighbors: [
        {
          value: 6,
          neighbors: [
            {
              value: 7,
              neighbors: [],
            },
          ],
        },
      ],
    },
  ],
};
const res = BFS(root);
console.log('layer', res);
